src/ucp/core/ucp_am.c

```diff
@@ -0,0 +1,564 @@
+     * If it is completed immediately, release the request and return the status.
+     * Otherwise, return the request.
+     */
+
+    status = ucp_request_send(req, 0);
+    if (req->flags & UCP_REQUEST_FLAG_COMPLETED) {
+        ucs_trace_req("releasing send request %p, returning status %s", req,
+                      ucs_status_string(status));
+        ucp_request_put(req);
+        return UCS_STATUS_PTR(status);
+    }
+
+    ucp_request_set_callback(req, send.cb, cb);
+    
+    return req + 1;
+}
+
+ucs_status_ptr_t ucp_am_send_nb(ucp_ep_h ep, uint16_t id,
+                                const void *payload, size_t count,
+                                uintptr_t datatype, ucp_send_callback_t cb,
+                                unsigned flags)
+{
+    ucs_status_t status;
+    ucs_status_ptr_t ret;
+    ucp_request_t *req;
+    size_t length;
+    
+    if (ucs_unlikely((flags != 0) && !(flags & UCP_AM_SEND_REPLY))) {
+        return UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
+    }
+
+    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
+
+
+    status = ucp_ep_resolve_dest_ep_ptr(ep, ep->am_lane);
+    if (status != UCS_OK) {
+        ret = UCS_STATUS_PTR(status);
+        goto out;
+    }
+    if (!(flags & UCP_AM_SEND_REPLY) && 
+        (ucs_likely(UCP_DT_IS_CONTIG(datatype)))) {
+        length = ucp_contig_dt_length(datatype, count);
+        
+        if (ucs_likely((ssize_t)length <= ucp_ep_config(ep)->am.max_short)) {
+            status = ucp_am_send_short(ep, id, payload, length);
+            if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
+                UCP_EP_STAT_TAG_OP(ep, EAGER);
+                ret = UCS_STATUS_PTR(status);
+                goto out;
+            }
+        }
+    }
+    
+    req = ucp_request_get(ep->worker);
+    if (ucs_unlikely(req == NULL)) {
+        ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+        goto out;
+    }
+
+    ucp_am_send_req_init(req, ep, payload, datatype, count, flags, id);
+  
+    ret = ucp_am_send_req(req, count, &ucp_ep_config(ep)->am, cb,
+                          ucp_ep_config(ep)->am_u.proto);
+
+out:
+    UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(ep->worker);
+    return ret;
+}
+
+static ucs_status_t 
+ucp_am_handler(void *am_arg, void *am_data, size_t am_length,
+               unsigned am_flags)
+{
+    ucp_worker_h worker = (ucp_worker_h) am_arg;
+    ucp_am_hdr_t *hdr = (ucp_am_hdr_t *) am_data;
+    ucp_recv_desc_t *desc = NULL;
+    uint16_t recv_flags = 0;
+    uint16_t am_id      = hdr->am_id;
+    ucs_status_t status;
+    ucp_am_reply_hdr_t *reply_hdr;
+    ucp_ep_h reply_ep;
+    size_t hdr_size;
+
+    if (hdr->am_id > worker->am_cb_array_len) {
+        ucs_warn("UCP Active Message was received with id : %u, but there" 
+                 "is no registered callback for that id", hdr->am_id);
+        return UCS_OK;
+    }
+    if (worker->am_cbs[hdr->am_id].cb == NULL) {
+        ucs_warn("UCP Active Message was received with id : %u, but there" 
```

Comment:
I am worried than over indexing the array would cause a seg fault. I know that if I put the length check first in the "if" statement that it would likely do the length check first, but I was unsure if thats guaranteed.

Reply:
ordering is guaranteed by C-standard, otherwise tons of code like
```
if (ptr && ptr->field) {
...
}
```
wouldn't work :) the same about `||`

Reply:
Okay, thank you for the clarification